Mathlib/Data/ListM/Basic.lean

```diff
@@ -361,6 +361,82 @@ This will run forever if the list is infinite. -/
 unsafe def fold (f : β → α → β) (init : β) (L : ListM m α) : m β :=
   L.foldM (fun b a => pure (f b a)) init
 
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+partial def getUpToFirstM (L : ListM m α) (p : α → m (ULift Bool)) : m (List α × ListM m α) := do
+  match ← L.uncons with
+  | none => return ([], nil)
+  | some (x, xs) => (if (← p x).down then
+      return ([x], xs)
+    else do
+      let (acc, R) ← getUpToFirstM xs p
+      return (x :: acc, R))
+
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+def getUpToFirst (L : ListM m α) (p : α → Bool) : m (List α × ListM m α) :=
+  L.getUpToFirstM fun a => pure (.up (p a))
+
+/--
+Extract a maximal prefix of a lazy list consisting of elements
+satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+
+(Note that the first element *not* satisfying the predicate will be generated,
+and pushed back on to the remaining lazy list.)
+-/
+partial def getWhileM (L : ListM m α) (p : α → m (ULift Bool)) :
+    m (List α × ListM m α) := do
+  match ← L.uncons with
+  | none => return ([], nil)
+  | some (x, xs) => (if (← p x).down then do
+      let (acc, R) ← getWhileM xs p
+      return (x :: acc, R)
+    else
+      return ([], cons do pure (some x, xs)))
+
+/--
+Extract a maximal prefix of a lazy list consisting of elements
+satisfying a predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+
+(Note that the first element *not* satisfying the predicate will be generated,
+and pushed back on to the remaining lazy list.)
+-/
+def getWhile (L : ListM m α) (p : α → Bool) : m (List α × ListM m α) :=
+  L.getWhileM fun a => pure (.up (p a))
+
+/--
+Partition a lazy list in runs of elements satisfying and not satisfying
+a monadic predicate.
+Return a lazy list of pairs of `List`s,
+where the first list consists of elements satisfying the predicate,
+and the second list consists of those that don't.
+-/
+partial def partitionM (L : ListM m α) (p : α → m (ULift Bool)) : ListM m (List α × List α) :=
```

Comment:
This API strikes me as slightly strange; why would I expect the yes and no lists to come in pairs? Wouldn't a `Sum (List α) (List α)` return type make more sense semantically?

Reply:
The function should have a different name as well, because it is different from `List.partition` or `List.partitionM`.

Reply:
I don't like `Sum (List α) (List α)` because that doesn't express that the lists will alternately be elements where `p` is true and where `p` is false. That is, in a `ListM m (Sum (List α) (List α))` you can have repeated `Sum.inl` or `Sum.inr`.

I was hoping that having the return type be a list of pairs expressed this alternating property.

My initial application for this is actually to chunk a lazy list of characters into "words with trailing whitespace". (i.e. I want to convert "  this is     nice" into ["  ", "this ", "is     ", "nice"]), and for that having the true sequences and false sequences paired up was very convenient.

I'll try to think of a different name...

Reply:
Maybe the underlying function here should be
```
def runsM (L : ListM m α) (f : α → m β) : ListM m (β × List α) := ...
```
which splits the list into runs of elements with the same value under `f`.

Reply:
Okay, I have dropped `partition` entirely, replacing it with `runs` and `splitAtBecomesTrue`, which are hopefully more general, with more understandable signatures, and solve my application better anyway.